libraries/AP_Scheduler/AP_Scheduler.h

```diff
@@ -45,27 +36,34 @@
   To run tasks use scheduler.run(), passing the amount of time that
   the scheduler is allowed to use before it must return
  */
+template <class Vehicle>
+struct AP_Task {
+// typedefs
+    typedef void (Vehicle::*task_fn_t)();
+// variables
+    task_fn_t   _function;
+    
+    const char* _name;
+    float       _rate_hz;
+    uint16_t    _max_time_micros;
+// statics
+    static AP_Task<Vehicle> create(task_fn_t function, float rate_hz, uint16_t max_time_micros) {
```

Comment:
This is looking good, but one thing you need to pay attention is that this method needs to be constexpr.  If you can't convince the compiler it's constexpr, then it's what would hold this PR (and not the use of template or removal of macros or anything like).  I think in c++11 you can make this constexpr since you have a single return statement (or was it actually added in c++14?)

Please check the generated object with objdump to see if the table of each vehicle is in a const section - otherwise for PX4 it will need to go to RAM instead of remaining in flash.